File: tracks/python/fundamentals/variables-and-types/variables.py
# =============================================================================
# Variables and Types — Python
# =============================================================================
# Fill in each section. Run with: python variables.py
# Or just save and ask for a review.
# =============================================================================

# =============================================================================
# 1. BASIC ASSIGNMENT
# Python has no declaration keyword. Assignment IS declaration.
# =============================================================================
# a) a string


# b) an int


# c) a float


# d) a bool


# =============================================================================
# 2. TYPE HINTS
# Python 3 supports type hints, but they're NOT enforced at runtime.
# They're for documentation and tools like mypy.
# =============================================================================
# a) name: str = "redis"


# b) port: int = 6379


# c) declare a variable with a WRONG type hint (e.g. x: int = "hello")
#    does Python complain? why or why not?


# =============================================================================
# 3. DYNAMIC TYPING
# A variable can be reassigned to a completely different type.
# =============================================================================
# a) assign x = 42, then x = "hello" — this is legal in Python


# b) use type() to check the type at each step
#    print(type(x))


# =============================================================================
# 4. NONE
# Python's equivalent of null/nil/undefined.
# =============================================================================
# a) assign a variable to None


# b) check its type with type()


# c) check if it's None using `is None` (not ==). Why `is` instead of `==`?
#    (answer in a comment)


# =============================================================================
# 5. MULTIPLE ASSIGNMENT
# Python supports several shorthand assignment forms.
# =============================================================================
# a) a, b, c = 1, 2, 3  (tuple unpacking)


# b) x = y = z = 0  (chained assignment)


# c) swap two variables without a temp: a, b = b, a


# =============================================================================
# 6. CONSTANTS (sort of)
# Python has no true constants. Convention is ALL_CAPS.
# =============================================================================
# a) MAX_RETRIES = 3


# b) try reassigning it — Python won't stop you. How does this compare
#    to Go's `const` or Rust's `const`? (answer in a comment)


# =============================================================================
# 7. PRINT EVERYTHING
# Use print() and f-strings. Use type() to show types.
# Example: print(f"port = {port} (type: {type(port).__name__})")
# =============================================================================

print("done")

File: tracks/java/fundamentals/variables-and-types/Variables.java
// ============================================================================
// Variables and Types — Java
// ============================================================================
// Fill in each section. Compile and run with:
//   javac Variables.java && java Variables
// Or just save and ask for a review.
// ============================================================================

public class Variables {
    public static void main(String[] args) {

        // ====================================================================
        // 1. PRIMITIVE TYPES
        // Java has 8 primitives. Declare one of each that you'll commonly use:
        // ====================================================================
        // a) int

        // b) double

        // c) boolean

        // d) char (single character — uses single quotes)

        // e) long (for large numbers — suffix with L)

        // ====================================================================
        // 2. REFERENCE TYPES (WRAPPER CLASSES)
        // Every primitive has a wrapper: int -> Integer, double -> Double, etc.
        // These are objects, not primitives. They can be null.
        // ====================================================================
        // a) Integer (can be null, unlike int)

        // b) String (always a reference type — no primitive for strings)

        // c) Boolean

        // ====================================================================
        // 3. VAR (TYPE INFERENCE)
        // Since Java 10, `var` lets the compiler infer the type.
        // Only works for local variables.
        // ====================================================================
        // a) var name = "redis";

        // b) var port = 6379; — what type does Java infer?

        // c) can you use var without an initializer? (try it, answer in comment)

        // ====================================================================
        // 4. FINAL (CONSTANTS)
        // `final` prevents reassignment — similar to const in other languages.
        // ====================================================================
        // a) final int maxRetries = 3;

        // b) try reassigning it — what error do you get?

        // c) final String name = "redis"; then try name = "kafka";

        // ====================================================================
        // 5. DEFAULT VALUES vs COMPILE ERRORS
        // Class-level fields get defaults (0, false, null).
        // Local variables do NOT — you must initialize before use.
        // ====================================================================
        // a) declare a local int without initializing, then try to print it.
        //    what happens?

        // b) how does this compare to Go's zero values? (answer in a comment)

        // ====================================================================
        // 6. PRINT EVERYTHING
        // Use System.out.println or System.out.printf.
        // To check types: use .getClass().getSimpleName() on wrapper types.
        // For primitives, you can cast to Object first or just state the type.
        // Example: System.out.printf("port = %d (int)%n", port);
        // ====================================================================

        System.out.println("done");
    }
}
